lib/route/src/Obelisk/Route.hs

```diff
@@ -927,19 +930,45 @@ pageNameEncoder = bimap
   (unpackTextEncoder . prefixTextEncoder "/" . pathSegmentsTextEncoder . listToNonEmptyEncoder)
   (unpackTextEncoder . prefixNonemptyTextEncoder "?" . queryParametersTextEncoder . toListMapEncoder)
 
+--------------------------------------------------------------------------------
+-- Handle decoding errors
+--------------------------------------------------------------------------------
 -- | Handle an error in parsing, for example, in order to redirect to a 404 page.
 handleEncoder
-  :: (Functor check)
+  :: Functor check
   => (e -> a)
   -> Encoder check (Either e) a b
   -> Encoder check Identity a b
-handleEncoder recover e = Encoder $ do
-  i <- unEncoder e
-  return $ i
-    { _encoderImpl_decode = \a -> pure $ case _encoderImpl_decode i a of
-      Right r -> r
-      Left err -> recover err
-    }
+handleEncoder recover = unsafeWrapDecoding (either (Identity . recover) pure .)
+
+-- | When parsing fails, try to decode as another element.
+-- Useful for canonicalizing on decode, e.g. trimming whitespace, rounding, error margins.
+fallbackEncoder
+  :: (Functor check, MonadError err parse)
+  => (err -> b -> b)
+  -> Encoder check parse a b
+  -> Encoder check parse a b
+fallbackEncoder fallback = retryEncoder $ \e p -> p . fallback e
+
+-- | When parsing fails, try again with the provided function.
+retryEncoder
+  :: (Functor check, MonadError err parse)
+  => (err -> (b -> parse a) -> (b -> parse a))
+  -> Encoder check parse a b
+  -> Encoder check parse a b
+retryEncoder retry = unsafeWrapDecoding $ \p b -> catchError (p b) $ \e -> retry e p b
```

Comment:
Note this only satisfies `unsafeWrapDecoding`'s law if we can trust MonadError to preserve success. Which I wager we should, but the documentation has no laws whatsoever.
Currently being discussed in the mailing list: https://mail.haskell.org/pipermail/libraries/2022-September/031621.html

Reply:
Good enough